src/main/scala/nl/knaw/dans/api/sword2/DepositHandler.scala

```diff
@@ -152,54 +166,157 @@ object DepositHandler {
+          if (file.exists()) {
+            Failure(InvalidDepositException(id, s"File ${item.getFilename} in the fetch.txt is already present in the bag."))
+          } else {
+            file.getParentFile.mkdirs()
+            Files.copy(src, file.toPath)
+          }
+        }))
+    }
+  }
+
+  private def noFetchItemsAlreadyInBag(bagitDir: File, fetchItems: Seq[FetchTxt.FilenameSizeUrl])(implicit id: String): Try[Unit] = {
+    log.debug(s"[$id] Checking that the files in fetch.txt are absent in the bag.")
+
+    val presentFiles = fetchItems.filter(item => new File(bagitDir.getAbsoluteFile, item.getFilename).exists())
+    if (presentFiles.nonEmpty)
+      Failure(InvalidDepositException(id, s"Fetch.txt file ${presentFiles.head.getFilename} is already present in the bag."))
+    else
+      Success(Unit)
+  }
+
+  private def validateChecksumsFetchItems(bag: Bag, fetchItems: Seq[FetchTxt.FilenameSizeUrl]) (implicit id: String, baseDir: File, baseUrl: URI): Try[Unit] = {
+    log.debug(s"[$id] Validating checksums of those files in fetch.txt, that refer to the bag store.")
+
+    val fetchItemFiles = fetchItems.map(_.getFilename)
+    val checksums = bag.getPayloadManifests.asScala.toList.
+                      flatMap(_.asScala).
+                      filter({case (file, checksum) =>
+                        fetchItemFiles.contains(file)
+                      })
+    val urls = fetchItems.map(file => file.getFilename -> file.getUrl).toMap
+
+    val checksumMapping: List[(String, String, String)] =
+      checksums.map { case (file, checksum) =>
+        if (urls isDefinedAt file)
+          (file, checksum, urls(file))
+        else
+          throw InvalidDepositException(id, s"Checksum validation failed: missing Payload Manifest file $file not found in the fetch.txt.")
+      }
+
+    validateChecksums(checksumMapping)
   }
 
-  private def getBagFromDir(dir: File)(implicit id: String): Bag = {
+  private def validateChecksums(checksumMapping: List[(String, String, String)])(implicit id: String, baseDir: File, baseUrl: URI): Try[Unit] = {
+    val errors = checksumMapping.map({case (file, checksum, url) =>
+        compareChecksumAgainstReferredBag(file, checksum, url)
+      }
+    ).collectResults.get.filter(_!= null)
+
+    if (errors.nonEmpty){
+      val errorMessage = errors.map(_.getMessage).mkString
+      Failure(InvalidDepositException(id, errorMessage))
+    }
+    else
+      Success(Unit)
+  }
+
+  private def compareChecksumAgainstReferredBag(file: String, checksum: String, url: String) (implicit id: String, baseDir: File, baseUrl: URI): Try[Exception] = Try {
+    val referredFile = getReferredFile(url, baseUrl)
+    val referredBagChecksums = getReferredBagChecksums(url)
+    if (referredBagChecksums.contains(referredFile -> checksum))
+      null
+    else {
+      if (referredBagChecksums.map({case (file, _) => file}).contains(referredFile))
```

Comment:
According to my IDEA there is an issue with variable shadowing. You have a `file` as parameter of `compareChecksumAgainstReferredBag` and also have a `file` in this lambda.